src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java

```diff
@@ -1,29 +1,132 @@
 package org.zalando.nakadi.repository.kafka;
 
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.clients.producer.KafkaProducer;
 import org.apache.kafka.clients.producer.Producer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.annotation.Profile;
 import org.springframework.stereotype.Component;
 
-import java.util.List;
-
 @Component
 @Profile("!test")
 public class KafkaFactory {
+
     private final KafkaLocationManager kafkaLocationManager;
-    private final KafkaProducer<String, String> kafkaProducer;
+    private final Counter useCountMetric;
+    private final Counter producerTerminations;
+    @Nullable
+    private Producer<String, String> activeProducer;
+    private final Map<Producer<String, String>, AtomicInteger> useCount = new ConcurrentHashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaFactory.class);
 
     @Autowired
-    public KafkaFactory(final KafkaLocationManager kafkaLocationManager) {
+    public KafkaFactory(final KafkaLocationManager kafkaLocationManager, final MetricRegistry metricRegistry) {
         this.kafkaLocationManager = kafkaLocationManager;
-        kafkaProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
+        this.useCountMetric = metricRegistry.counter("kafka.producer.use_count");
+        this.producerTerminations = metricRegistry.counter("kafka.producer.termination_count");
+    }
+
+    @Nullable
+    private Producer<String, String> takeUnderLock(final boolean canCreate) {
+        final Lock lock = canCreate ? rwLock.writeLock() : rwLock.readLock();
+        lock.lock();
+        try {
+            if (null != activeProducer) {
+                useCountMetric.inc();
+                useCount.get(activeProducer).incrementAndGet();
+                return activeProducer;
+            } else if (canCreate) {
+                activeProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
+                useCountMetric.inc();
+                useCount.put(activeProducer, new AtomicInteger(1));
+                LOG.info("New producer instance created: " + activeProducer);
+                return activeProducer;
+            } else {
+                return null;
```

Comment:
Because in first case it takes readLock (e.g. many threads can call this method at the same time). Second call is under write lock, which means, that only one thread is creating this producer. Thus, I guarantee, that I will now spawn several producers at the same time